Documentation/proposals/app-level-api.md

```diff
@@ -0,0 +1,234 @@
+# Fine-Grained Pod Manipulations
+
+To provide and API for imperative application operations (e.g. start|stop|add)
+inside a pod for finer-grained control over rkt containerization concepts and
+debugging needs, this proposal introduces new stage1 entrypoints and a
+subcommand CLI API that will be used for manipulating applications inside pods.
+
+The motivation behind this change is the direction orchestration systems are
+taking and to facilitate this new direction. For more details, see
+[kubernetes#25899](https://github.com/yujuhong/kubernetes/blob/08dc66113399c89e31f6872f3c638695a6ec6a8d/docs/proposals/container-runtime-interface-v1.md).
+
+# API
+
+The envisioned workflow for the app-level CLI API is that after a pod has been started, 
+the users will be invoking the rkt CLI which consist of application logic on top of aforementioned stage1 entrypoints.
+
+## rkt app add
+Injects an application image into a running pod.
+
+It first prepares an application rootfs in stage0 for the application image/images
+and then the prepared-app is injected via the entrypoint.
+
+```bash
+rkt app add <pod-uuid> --app=<app-name> <image-name/hash/address/registry-URL> <arguments>
+```
+
+**Note:** Not every pod will be injectable and it will be enabled through an option.
+
+## rkt app rm
+Removes an application image into a running pod, meaning that the leftover resources are removed from the pod.
+
+```bash
+rkt app rm <pod-uuid> --app=<app-name> <arguments>
+```
+
+**Note:** when a pod becomes empty, currently it will terminate, but this
+proposal will introduce a `--mutable` or `--allow-empty` or `--dumb` flag to be
+used when starting pods, so that the lifecycle management of the pod is left to
+the user. Meaning that empty pods won't be terminated.
+
+### Resources Leftover by an Application in Default Stage1 Flavor
+- Rootfs (e.g. `/opt/stage/<app-name>`)
+- Mounts from volumes (e.g. `/opt/stage/<app-name>/<volume-name>`)
+- Mounts related to rkt operations (e.g. `/opt/stage/<app-name>/dev/null`)
+- Systemd service files (e.g. `<app-name>.service` and `reaper-<app-name>.service`)
+- Miscellaneous files (e.g. `/rkt/<app-name>.env`, `/rkt/status...`)
+
+## rkt app start
+Starts an application that was injected.
+This operation is idempotent.
+
+```bash
+rkt app start <pod-uuid> --app=<app-name> <arguments>
+```
+
+## rkt app stop
+Stops a running application gracefully (grace is defined in the `app stop` entrypoint section).
+
+```bash
+rkt app stop <pod-uuid> --app=<app-name>
+```
+
+## rkt app list
+Lists the applications that are inside a pod, running or stopped.
+
+```bash
+rkt app list <pod-uuid> <arguments>
+```
+
+**Note:** List should consist of an app specifier and status at the very least,
+the rest is up for discussions.
+
+## rkt app status
+Returns the execution status of application inside a pod.
+
+```bash
+rkt app status <pod-uuid> --app=<app-name> <arguments>
+```
```

Comment:
I don't see the interface for getting the app status. Now the current [stage1 implementors guide]https://github.com/coreos/rkt/blob/master/Documentation/devel/stage1-implementors-guide.md) defines the file to save the [app exit status](https://github.com/coreos/rkt/blob/master/Documentation/devel/stage1-implementors-guide.md#status) since there isn't the AppStatus concept.